### PR TITLE
impr(result): display input history heatmap in the selected speed unit

### LINKED ISSUE #4531: Input history heatmap is not using the new speed unit setting

### PR DESCRIPTION
**Pull request by ridemountainpig**
### Description
Input history heatmap is add the new speed unit setting

wpm:

https://github.com/monkeytypegame/monkeytype/assets/92412722/03525c64-0543-4fc5-aeec-696cef589781

cpm:

https://github.com/monkeytypegame/monkeytype/assets/92412722/defc05c4-594c-4711-a5e7-8ee39a93aed5

wps:

https://github.com/monkeytypegame/monkeytype/assets/92412722/ccff5c0c-e590-46e0-864b-120f9e410573

cps:

https://github.com/monkeytypegame/monkeytype/assets/92412722/fee0261e-ec9f-4594-bb08-0302f9424846

Closes #4531

### CODE REVIEW

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/5862498133

**Review by Pranay-Kamble — APPROVED** [review]

File: frontend/src/ts/controllers/input-controller.ts [hunk 1]
```diff
@@ -179,7 +180,9 @@ function handleSpace(): void {
 
   const burst: number = TestStats.calculateBurst();
   LiveBurst.update(Math.round(burst));
-  TestInput.pushBurstToHistory(burst);
+  TestInput.pushBurstToHistory(
```
Comment (fehmer) [inline frontend/src/ts/controllers/input-controller.ts:183]: Should we keep the data in wpm and only convert to the user defined unit when showing it to the user? The speed unit can be changed on the result screen as well.

Comment (ridemountainpig) [reply frontend/src/ts/controllers/input-controller.ts:183]: I think only changed speed unit on the result screen sound great, @Miodec how do you think, which way is better?

Comment (ridemountainpig) [reply frontend/src/ts/controllers/input-controller.ts:183]: I think change the data push to burst history is more better, if didn't change burst history just change speed unit on the result screen, there will have many `TestInput.burstHistory` in the code need to change also.

CleanShot 2023-08-15 at 16 28 25@2x

Comment (fehmer) [reply frontend/src/ts/controllers/input-controller.ts:183]: does the current solution supports switching the unit on the result screen p

Comment (ridemountainpig) [reply frontend/src/ts/controllers/input-controller.ts]: > Should we keep the data in wpm and only convert to the user defined unit when showing it to the user? The speed unit can be changed on the result screen as well.

@fehmer I have change the code to your provide solution, and the heatmap will change when switching the unit on the result screen using the command palette now.

https://github.com/monkeytypegame/monkeytype/assets/92412722/c6646c59-6671-454e-b0f6-3653d9eed641

File: frontend/src/ts/config.ts [hunk 2]
```diff
@@ -399,6 +400,7 @@ export function setTypingSpeedUnit(
   config.typingSpeedUnit = val;
   saveToLocalStorage("typingSpeedUnit", nosave);
   ConfigEvent.dispatch("typingSpeedUnit", config.typingSpeedUnit, nosave);
+  applyBurstHeatmap();
```
Comment (fehmer) [inline frontend/src/ts/config.ts:403]: We should not call it here. There is an event triggered when updating the typingSpeedUnit and you can listen on that event on the result screen

Comment (fehmer) [inline frontend/src/ts/config.ts:403]: We should not call it here. There is an event triggered when updating the typingSpeedUnit and you can listen on that event on the result screen

**Review by fehmer — CHANGES_REQUESTED** [review]

File: frontend/src/ts/config.ts (hunk 2 above)
Comment (ridemountainpig) [reply frontend/src/ts/config.ts]: I have update it in new commit.

**Comment by fehmer** [comment]
display for wps causes problems:
image

We could use steps of `0.5` as we do with the histogram chart on the accounts page.

**Comment by ridemountainpig** [comment]
CleanShot 2023-08-17 at 14 40 22@2x

I think it look good right now.

**Comment by Miodec** [comment]
> display for wps causes problems: image
> 
> We could use steps of `0.5` as we do with the histogram chart on the accounts page.

Idk, even with 0.5 there just isnt enough resolution, just the drawback of wps.

File: frontend/src/ts/test/test-ui.ts [hunk 3]
```diff
@@ -1080,6 +1080,12 @@ export async function applyBurstHeatmap(): Promise<void> {
     burstlist = burstlist.filter((x) => x !== Infinity);
     burstlist = burstlist.filter((x) => x < 350);
 
+    burstlist.forEach((burst, index) => {
+      burstlist[index] = Math.round(
+        getTypingSpeedUnit(Config.typingSpeedUnit).fromWpm(burst)
```
Comment (Miodec) [inline frontend/src/ts/test/test-ui.ts:1085]: Maybe extract the speed unit outside of the loop? Then only call `fromWpm` inside the loop.

File: frontend/src/ts/test/result.ts [hunk 4]
```diff
@@ -888,6 +888,7 @@ ConfigEvent.subscribe(async (eventKey) => {
     updateWpmAndAcc();
     await updateGraph();
     await updateGraphPBLine();
+    TestUI.applyBurstHeatmap();
```
Comment (Miodec) [inline frontend/src/ts/test/result.ts:891]: Will this apply the heatmap even if its disabled? Or only when its visible?

File: frontend/src/ts/test/test-ui.ts [hunk 5]
```diff
@@ -1165,7 +1171,10 @@ export async function applyBurstHeatmap(): Promise<void> {
       if (wordBurstAttr === undefined) {
         $(word).css("color", unreachedColor);
       } else {
-        const wordBurstVal = parseInt(<string>wordBurstAttr);
+        let wordBurstVal = parseInt(<string>wordBurstAttr);
```
Comment (Miodec) [inline frontend/src/ts/test/test-ui.ts:1174]: Could you just change `<string>` to `to string`? We don't really use this syntax often. (But i know its not your change)

**Review by Miodec — CHANGES_REQUESTED** [review]

File: frontend/src/ts/test/result.ts (hunk 4 above)
Comment (ridemountainpig) [reply frontend/src/ts/test/result.ts:891]: It will run when user change speed unit.

Comment (Miodec) [reply frontend/src/ts/test/result.ts:891]: Yes, i know. But what if the heatmap is not enabled? Will this automatically enable it?

Comment (ridemountainpig) [reply frontend/src/ts/test/result.ts:891]: It will only run when user change speed unit in after test result page, in other page won't.

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/5890394588

**Pull request merged**